File: crates/vici/src/command.rs
//! The command vocabulary: what a resolved key sequence means.
//!
//! Everything here is plain data — no closures, no function pointers — so a
//! keymap can be deserialised from config later without redesign.

use crate::motion::Find;

/// Editing mode.
///
/// Operator-pending is deliberately *not* a mode. It is transient parser state
/// ([`crate::Pending`]), which keeps the mode enum small and stops "am I in
/// operator-pending?" from leaking into rendering and keymap lookup.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
pub enum Mode {
    #[default]
    Normal,
    Insert,
    Replace,
    Visual(VisualKind),
}

impl Mode {
    /// True where the count/operator/motion grammar applies.
    #[must_use]
    pub const fn is_command(self) -> bool {
        matches!(self, Self::Normal | Self::Visual(_))
    }

    #[must_use]
    pub const fn is_visual(self) -> bool {
        matches!(self, Self::Visual(_))
    }
}

/// Visual-mode granularity. Block mode is not implemented.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum VisualKind {
    Char,
    Line,
}

/// A cursor movement.
///
/// Each motion carries its own operator semantics via [`Motion::is_linewise`] and
/// [`Motion::is_inclusive`]. Getting these right is what makes `dw` and `de`
/// differ by one character, and `dj` delete two whole rows.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Motion {
    Left,
    Right,
    Down,
    Up,
    /// `0`
    FirstColumn,
    /// `^`
    FirstNonBlank,
    /// `$`
    LastColumn,
    /// `w` / `W`
    WordForward {
        big: bool,
    },
    /// `b` / `B`
    WordBackward {
        big: bool,
    },
    /// `e` / `E`
    WordEnd {
        big: bool,
    },
    /// `f` / `F` / `t` / `T`
    Find(Find),
    /// `;` / `,`
    RepeatFind {
        reverse: bool,
    },
    /// `G` — count as an absolute row, else the last row.
    GotoRow,
    /// `gg` — count as an absolute row, else the first row.
    GotoFirstRow,
    /// `%` — the match of the first bracket on the row, or of a quote.
    MatchPair,
    /// `H` — a counted row from the top of the reported screen.
    ScreenTop,
    /// `M` — the middle row of the reported screen.
    ScreenMiddle,
    /// `L` — a counted row from the bottom of the reported screen.
    ScreenBottom,
}

impl Motion {
    /// Whether an operator over this motion acts on whole rows.
    #[must_use]
    pub const fn is_linewise(self) -> bool {
        matches!(
            self,
            Self::Down
                | Self::Up
                | Self::GotoRow
                | Self::GotoFirstRow
                | Self::ScreenTop
                | Self::ScreenMiddle
                | Self::ScreenBottom
        )
    }

    /// Whether the character under the motion's destination is included.
    #[must_use]
    pub const fn is_inclusive(self) -> bool {
        match self {
            // `%` takes the delimiter it lands on with it, in either direction:
            // `d%` from either end of a pair deletes both and everything between.
            Self::WordEnd { .. } | Self::LastColumn | Self::MatchPair => true,
            // Forward `f` and `t` are both inclusive; `t` simply lands a character
            // earlier. So `dt,` deletes up to but not including the comma, which
            // requires including the character `t` landed on. Backward `F`/`T` are
            // exclusive, leaving the character under the cursor alone.
            Self::Find(Find { backward, .. }) => !backward,
            _ => false,
        }
    }
}

/// A text object's extent.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum ObjectScope {
    /// `i` — contents only.
    Inner,
    /// `a` — contents plus delimiters or trailing whitespace.
    Around,
}

/// A structural region, selected by `i`/`a` plus a key.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TextObject {
    Word {
        big: bool,
    },
    /// A paired delimiter, e.g. `(`/`)` for `ib`.
    Delimited {
        open: char,
        close: char,
    },
    /// A symmetric delimiter, e.g. `"` or `` ` ``.
    Quoted(char),
    Paragraph,
}

/// What an operator acts upon.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Target {
    Motion(Motion),
    Object {
        scope: ObjectScope,
        object: TextObject,
    },
    /// A doubled operator: `dd`, `cc`, `yy`.
    CurrentRow,
    /// The active visual selection.
    Selection,
}

/// An operator, awaiting a [`Target`].
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Operator {
    Delete,
    Change,
    Yank,
    /// `>`
    ShiftRight,
    /// `<`
    ShiftLeft,
    /// `gu`
    Lower,
    /// `gU`
    Upper,
    /// `g~`, and `~` over a visual selection.
    SwapCase,
}

impl Operator {
    /// Whether this operator fills the register.
    ///
    /// Case-changing and shift operators do not: `gUW` or `>>` leave whatever
    /// you last yanked alone, so a `p` afterwards still pastes it.
    #[must_use]
    pub const fn yanks(self) -> bool {
        matches!(self, Self::Delete | Self::Change | Self::Yank)
    }

    /// Whether this operator widens every target to complete rows.
    ///
    /// Shifting is linewise even over a characterwise motion or visual
    /// selection: `>w` shifts the row containing the word, not just the word.
    #[must_use]
    pub const fn forces_linewise(self) -> bool {
        matches!(self, Self::ShiftRight | Self::ShiftLeft)
    }
}

/// Where insert mode begins relative to the cursor.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum InsertAt {
    /// `i`
    Cursor,
    /// `a`
    After,
    /// `I`
    FirstNonBlank,
    /// `A`
    EndOfRow,
    /// `o`
    RowBelow,
    /// `O`
    RowAbove,
}

/// Viewport movements. The core does not own the viewport — these resolve to
/// effects the host fulfils.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum Scroll {
    HalfPageDown,
    HalfPageUp,
    PageDown,
    PageUp,
    /// `zz`
    Center,
    /// `zt`
    Top,
    /// `zb`
    Bottom,
}

/// A fully resolved command.
///
/// Counts are *not* carried here — see [`crate::Resolution::Command`]. Keeping
/// them separate means the same `Command` value can sit in a keymap unchanged.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Command {
    Move(Motion),
    Operate {
        operator: Operator,
        target: Target,
    },
    /// Extend the visual selection to a text object.
    SelectObject {
        scope: ObjectScope,
        object: TextObject,
    },

    EnterInsert(InsertAt),
    EnterVisual(VisualKind),
    EnterReplace,
    /// `<Esc>` from any mode.
    EnterNormal,

    /// `x` / `X`
    DeleteChar {
        before: bool,
    },
    /// `r{char}`
    ReplaceChar(char),
    /// `J`
    JoinRows,
    /// `p` / `P`
    Put {
        before: bool,
    },
    /// `~`
    SwapCase,

    Undo,
    Redo,
    /// `.`
    Repeat,

    /// `q{register}` — a second `q` stops recording.
    RecordMacro(char),
    /// `@{register}`
    PlayMacro(char),

    Scroll(Scroll),

    /// `:` — the host opens a prompt.
    CommandPrompt,

    /// Literal text typed in insert mode.
    InsertText(char),
    /// `<CR>` in insert mode.
    InsertNewline,
    /// `<BS>` in insert mode.
    DeleteBack,
    /// `<C-w>` in insert mode.
    DeleteWordBack,
}

/// A binding that needs one further key as a literal character argument.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub enum AwaitChar {
    /// `f` / `F` / `t` / `T`
    Find { backward: bool, till: bool },
    /// `r`
    ReplaceChar,
    /// `q`
    RecordMacro,
    /// `@`
    PlayMacro,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn operator_semantics_of_motions() {
        // `dw` stops before the next word; `de` eats the last character.
        assert!(!Motion::WordForward { big: false }.is_inclusive());
        assert!(Motion::WordEnd { big: false }.is_inclusive());

        // `dj` takes whole rows.
        assert!(Motion::Down.is_linewise());
        assert!(!Motion::Right.is_linewise());

        // Both forward finds are inclusive of where they land: `t` lands one
        // character earlier, which is how `dt,` stops before the comma.
        let forward = |till| {
            Motion::Find(Find {
                target: ',',
                backward: false,
                till,
            })
        };
        assert!(forward(false).is_inclusive());
        assert!(forward(true).is_inclusive());
        // Backward finds leave the character under the cursor alone.
        assert!(
            !Motion::Find(Find {
                target: ',',
                backward: true,
                till: false
            })
            .is_inclusive()
        );
    }

    #[test]
    fn operator_pending_is_not_a_mode() {
        assert!(Mode::Normal.is_command());
        assert!(Mode::Visual(VisualKind::Line).is_command());
        assert!(!Mode::Insert.is_command());
        assert!(Mode::Visual(VisualKind::Char).is_visual());
    }
}

File: js/src/command.js
// The command vocabulary: what a resolved key sequence means.
//
// Everything here is plain data — small integers and frozen records, no
// closures — so a keymap can be deserialised from config later without
// redesign, and so the reducer's dispatch stays a dense switch.

/** Editing mode. Operator-pending is deliberately *not* a mode: it is parser state. */
export const NORMAL = 0;
export const INSERT = 1;
export const REPLACE = 2;
export const VISUAL = 3;
export const VISUAL_LINE = 4;

/** @param {number} mode @returns {boolean} True where the count/operator/motion grammar applies. */
export const isCommandMode = (mode) => mode === NORMAL || mode >= VISUAL;
/** @param {number} mode @returns {boolean} */
export const isVisual = (mode) => mode >= VISUAL;

/** Operators, awaiting a target. */
export const DELETE = 0;
export const CHANGE = 1;
export const YANK = 2;
export const SHIFT_RIGHT = 3;
export const SHIFT_LEFT = 4;
export const LOWER = 5;
export const UPPER = 6;
export const SWAP = 7;

/**
 * Whether an operator fills the register. Case-changing and shift operators do
 * not: `gUW` or `>>` leave whatever you last yanked alone.
 * @param {number} operator
 * @returns {boolean}
 */
export const yanks = (operator) => operator <= YANK;

/**
 * Whether an operator widens every target to complete rows. Shifting is
 * linewise even over a characterwise motion: `>w` shifts the row.
 * @param {number} operator
 * @returns {boolean}
 */
export const forcesLinewise = (operator) => operator === SHIFT_RIGHT || operator === SHIFT_LEFT;

/** What an operator acts upon. */
export const T_MOTION = 0;
export const T_OBJECT = 1;
/** A doubled operator: `dd`, `cc`, `yy`. */
export const T_CURRENT_ROW = 2;
/** The active visual selection. */
export const T_SELECTION = 3;

/** Where insert mode begins relative to the cursor. */
export const AT_CURSOR = 0;
export const AT_AFTER = 1;
export const AT_FIRST_NON_BLANK = 2;
export const AT_END_OF_ROW = 3;
export const AT_ROW_BELOW = 4;
export const AT_ROW_ABOVE = 5;

/** Viewport movements. The core does not own the viewport. */
export const HALF_PAGE_DOWN = 0;
export const HALF_PAGE_UP = 1;
export const PAGE_DOWN = 2;
export const PAGE_UP = 3;
export const CENTER = 4;
export const TOP = 5;
export const BOTTOM = 6;

/** Commands. */
export const MOVE = 0;
export const OPERATE = 1;
export const SELECT_OBJECT = 2;
export const ENTER_INSERT = 3;
export const ENTER_VISUAL = 4;
export const ENTER_REPLACE = 5;
export const ENTER_NORMAL = 6;
export const DELETE_CHAR = 7;
export const REPLACE_CHAR = 8;
export const CHANGE_SURROUND = 9;
export const DELETE_SURROUND = 10;
export const SURROUND_SELECTION = 11;
export const JOIN_ROWS = 12;
export const PUT = 13;
export const SWAP_CASE = 14;
export const UNDO = 15;
export const REDO = 16;
export const REPEAT = 17;
export const RECORD_MACRO = 18;
export const PLAY_MACRO = 19;
export const SET_MARK = 20;
export const JUMP_BACK = 21;
export const JUMP_FORWARD = 22;
export const SCROLL = 23;
export const COMMAND_PROMPT = 24;
export const INSERT_TEXT = 25;
export const INSERT_NEWLINE = 26;
export const DELETE_BACK = 27;
export const DELETE_WORD_BACK = 28;

/** Bindings that need one further key as a literal character argument. */
export const AWAIT_FIND = 0;
export const AWAIT_REPLACE_CHAR = 1;
export const AWAIT_RECORD = 2;
export const AWAIT_PLAY = 3;
export const AWAIT_SET_MARK = 4;
export const AWAIT_GOTO_MARK = 5;
export const AWAIT_SURROUND_TARGET = 6;
export const AWAIT_SURROUND_TO = 7;
export const AWAIT_SURROUND_SELECTION = 8;

/** Binding kinds, as stored in a keymap. */
export const B_COMMAND = 0;
export const B_OPERATOR = 1;
export const B_MOTION = 2;
export const B_SCOPE = 3;
export const B_AWAIT = 4;
export const B_SEARCH = 5;
